aidial_adapter_openai/gpt4_vision.py

```diff
@@ -0,0 +1,385 @@
+            return
+
+        if "id" in chunk:
+            id = chunk["id"]
+
+        if "created" in chunk:
+            created = chunk["created"]
+
+        choice: dict = (chunk.get("choices") or [{}])[0]
+
+        finish_type: Optional[str] = (choice.get("finish_details") or {}).get(
+            "type"
+        )
+
+        match finish_type:
+            case None:
+                pass
+            case "stop":
+                finish_reason = "stop"
+            case "max_tokens":
+                finish_reason = "length"
+            case "content_filter":
+                finish_reason = "content_filter"
+            case _:
+                logger.warning(
+                    f"Unknown finish type: {finish_type}. Defaulting to stop"
+                )
+
+        content = choice.get("delta", {}).get("content")
+
+        if content is None:
+            continue
+
+        yield chunk_format(
+            build_chunk(id, None, {"content": content}, created, is_stream)
+        )
+
+    yield chunk_format(build_chunk(id, finish_reason, {}, created, is_stream))
+    yield END_CHUNK
+
+
+def derive_image_content_type(attachment: Any) -> Optional[str]:
+    type = attachment.get("type")
+    if type is None:
+        return None
+
+    if (
+        type.startswith("image/")
+        and type[len("image/") :] in SUPPORTED_IMAGE_TYPES
+    ):
+        return type
+
+    if "octet-stream" in type:
+        # It's an arbitrary binary file. Trying to guess the type from the URL.
+        url = attachment.get("url")
+        if url is None:
+            return None
+
+        file_ext = url.split(".")[-1]
+
+        if file_ext in SUPPORTED_IMAGE_TYPES:
+            return f"image/{file_ext}"
+
+    return None
+
+
+async def download_image_attachment(
+    file_storage: Optional[FileStorage], attachment: Any
+) -> Optional[ImageSubmessage]:
+    type = derive_image_content_type(attachment)
+    if type is None:
+        return None
+
+    if "data" in attachment:
+        url = base64_to_image_url(type, attachment["data"])
+        return create_image_message(url)
+
+    if "url" in attachment:
+        attachment_link: str = attachment["url"]
+
+        try:
+            if file_storage is not None:
+                url = file_storage.attachment_link_to_url(attachment_link)
+                base64_str = await file_storage.download_file_as_base64(url)
+            else:
+                base64_str = await download_file_as_base64(attachment_link)
+
+            image_url = base64_to_image_url(type, base64_str)
+            return create_image_message(image_url)
+        except Exception:
+            logger.warning(
+                f"Failed to download image from URL: {attachment_link}"
+            )
+            return create_image_message(attachment_link)
+
+    return None
+
+
+async def transform_message(
+    file_storage: Optional[FileStorage], message: dict
+) -> dict | str:
+    content = message.get("content", "")
```

Comment:
Let's change this to `message.get("content") or ""` because the content can be None when it has attachments

Reply:
Fixed